File: src/utils/webhook.ts
import { logger } from '~/config/logger';

interface IWebhookPayload {
  jobId: string;
  status: 'completed' | 'failed';
  result?: {
    url: string;
    fileSizeBytes: number;
    processingTimeMs: number;
  };
  error?: string;
  timestamp: string;
}

interface IBatchJobResult {
  jobId: string;
  model: string;
  status: 'completed' | 'failed';
  result?: {
    url: string;
    fileSizeBytes: number;
    processingTimeMs: number;
  };
  error?: string;
}

interface IBatchWebhookPayload {
  batchId: string;
  status: 'completed' | 'partial_failure';
  totalJobs: number;
  successfulJobs: number;
  failedJobs: number;
  results: IBatchJobResult[];
  timestamp: string;
}

export async function sendWebhook(
  webhookUrl: string,
  jobId: string,
  status: 'completed' | 'failed',
  result?: IWebhookPayload['result'],
  error?: string
): Promise<void> {
  const payload: IWebhookPayload = {
    jobId,
    status,
    result,
    error,
    timestamp: new Date().toISOString()
  };

  try {
    const response = await fetch(webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(10000) // 10s timeout
    });

    if (!response.ok) {
      logger.error({ jobId, webhookUrl, status: response.status }, 'Webhook delivery failed');
    } else {
      logger.info({ jobId, webhookUrl }, 'Webhook sent successfully');
    }
  } catch (err) {
    const errorMessage = err instanceof Error ? err.message : String(err);
    logger.error({ jobId, webhookUrl, error: errorMessage }, 'Webhook error');
  }
}

/**
 * Send batch completion webhook with aggregated results
 */
export async function sendBatchWebhook(
  webhookUrl: string,
  batchId: string,
  status: 'completed' | 'partial_failure',
  totalJobs: number,
  successfulJobs: number,
  failedJobs: number,
  results: IBatchJobResult[]
): Promise<void> {
  const payload: IBatchWebhookPayload = {
    batchId,
    status,
    totalJobs,
    successfulJobs,
    failedJobs,
    results,
    timestamp: new Date().toISOString()
  };

  try {
    const response = await fetch(webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(10000) // 10s timeout
    });

    if (!response.ok) {
      logger.error({ batchId, webhookUrl, status: response.status }, 'Batch webhook delivery failed');
    } else {
      logger.info({ batchId, webhookUrl, totalJobs, successfulJobs, failedJobs }, 'Batch webhook sent successfully');
    }
  } catch (err) {
    const errorMessage = err instanceof Error ? err.message : String(err);
    logger.error({ batchId, webhookUrl, error: errorMessage }, 'Batch webhook error');
  }
}

File: gpu-worker/worker/webhook.py
"""Webhook notifications for GPU workers."""
import logging
from datetime import datetime, timezone
from typing import Any

import requests

from .config import settings

logger = logging.getLogger(__name__)


def _call_webhook(url: str, payload: dict[str, Any]) -> bool:
    """Send a webhook notification.

    Args:
        url: Webhook URL
        payload: JSON payload to send

    Returns:
        True if successful, False otherwise
    """
    try:
        response = requests.post(
            url,
            json=payload,
            headers={
                "Content-Type": "application/json",
                "X-Webhook-Secret": settings.WEBHOOK_SECRET,
            },
            timeout=10,
        )
        response.raise_for_status()
        logger.info(f"Webhook sent successfully to {url}")
        return True
    except requests.RequestException as e:
        logger.error(f"Webhook failed to {url}: {e}")
        return False


def notify_complete(
    job_id: str,
    model: str,
    result: dict[str, Any],
    webhook_url: str | None = None,
) -> None:
    """Notify API and optional external webhook of job completion.

    Args:
        job_id: The job ID
        model: Model type (wav2lip, zimage, infinitetalk)
        result: Result dictionary with url, contentType, etc.
        webhook_url: Optional external webhook URL
    """
    payload = {
        "jobId": job_id,
        "status": "completed",
        "result": result,
        "processingTimeMs": result.get("processingTimeMs"),
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }

    # Notify the API server
    if settings.API_WEBHOOK_URL:
        _call_webhook(settings.API_WEBHOOK_URL, payload)

    # Notify external webhook if configured
    if webhook_url:
        _call_webhook(webhook_url, payload)


def notify_failed(
    job_id: str,
    model: str,
    error: str,
    webhook_url: str | None = None,
) -> None:
    """Notify API and optional external webhook of job failure.

    Args:
        job_id: The job ID
        model: Model type (wav2lip, zimage, infinitetalk)
        error: Error message
        webhook_url: Optional external webhook URL
    """
    payload = {
        "jobId": job_id,
        "status": "failed",
        "error": error,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }

    # Notify the API server
    if settings.API_WEBHOOK_URL:
        _call_webhook(settings.API_WEBHOOK_URL, payload)

    # Notify external webhook if configured
    if webhook_url:
        _call_webhook(webhook_url, payload)
